src/erlcloud_ec2.erl

```diff
@@ -1347,38 +1365,117 @@ attribute_xpath(block_device_mapping, AttributeName) ->
+    when is_list(Filter), is_integer(MaxResults) ->
+    describe_instances(Filter, MaxResults, undefined, default_config()).
+
+- spec describe_instances(ec2_instances_ids(), filter_list(), aws_config()) -> ec2_normal_describe_return();
+                         (filter_list(), ec2_max_result(), ec2_token()) -> ec2_paged_describe_return();
+                         (filter_list(), ec2_max_result(), aws_config()) -> ec2_paged_describe_return();
+                         (ec2_max_result(), ec2_token(), aws_config()) -> ec2_paged_describe_return().
 describe_instances(InstanceIDs, Filter, Config)
-  when is_list(InstanceIDs) ->
+    when is_list(InstanceIDs), is_list(Filter), is_record(Config, aws_config) ->
     Params = erlcloud_aws:param_list(InstanceIDs, "InstanceId") ++ list_to_ec2_filter(Filter),
     case ec2_query(Config, "DescribeInstances", Params, ?NEW_API_VERSION) of
         {ok, Doc} ->
             Reservations = xmerl_xpath:string("/DescribeInstancesResponse/reservationSet/item", Doc),
             {ok, [extract_reservation(Item) || Item <- Reservations]};
         {error, Reason} ->
             {error, Reason}
+    end;
+describe_instances(Filter, MaxResults, NextToken)
+    when is_list(Filter), is_integer(MaxResults), is_list(NextToken) ->
+    describe_instances(Filter, MaxResults, NextToken, default_config());
+describe_instances(Filter, MaxResults, Config)
+    when is_list(Filter), is_integer(MaxResults), MaxResults >= ?INSTANCES_MR_MIN, MaxResults =< ?INSTANCES_MR_MAX, is_record(Config, aws_config) ->
+    describe_instances(Filter, MaxResults, undefined, Config);
+describe_instances(MaxResults, NextToken, Config)
+    when is_integer(MaxResults), is_list(NextToken) orelse NextToken =:= undefined, is_record(Config, aws_config) ->
+    describe_instances([], MaxResults, NextToken, Config).
+
+-spec describe_instances(filter_list(), ec2_max_result(), ec2_token(), aws_config()) -> ec2_paged_describe_return().
+describe_instances(Filter, MaxResults, NextToken, Config)
+    when is_list(Filter), is_integer(MaxResults) andalso MaxResults >= ?INSTANCES_MR_MIN andalso MaxResults =< ?INSTANCES_MR_MAX, is_list(NextToken) orelse NextToken =:= undefined, is_record(Config, aws_config) ->
+    Params = [{"MaxResults", MaxResults}, {"NextToken", NextToken}] ++ list_to_ec2_filter(Filter),
+    case ec2_query(Config, "DescribeInstances", Params, ?NEW_API_VERSION) of
+        {ok, Doc} ->
+            Reservations = xmerl_xpath:string("/DescribeInstancesResponse/reservationSet/item", Doc),
+            NewNextToken = erlcloud_xml:get_text("/DescribeInstancesResponse/nextToken", Doc, undefined),
+            {ok, [extract_reservation(Item) || Item <- Reservations], NewNextToken};
+        {error, Reason} ->
+            {error, Reason}
     end.
 
+% Function for describing all instances page-by-page
+% Default page size is 1000
+-spec describe_instances_all() -> ec2_normal_describe_return().
+describe_instances_all() ->
```

Comment:
the point and the difference of *_all() is to provide a wrapper around paginated API to still get all the things but in chunks. 
this exist in other modules and added for consistency only. 
the might be unnecessary complication but well let's discuss.